README
-----------------------------------------------------------------------
Pre Make Kit README file

Document revision:
$Id$
-----------------------------------------------------------------------


Introduction
------------

The Pre Make Kit project is a toolkit providing a collection of tools
aimed to ease a lot of stages in project development. It provides a
scanner that generate template source components such as configuration
file or makefile. It also provides a dependency checker that sets up
templates files depending on the system components (such as auto* tools
or similar projects). For portability purpose, a BSD like install
program is also part of the kit.

The project aims the following goals :
- opensource (BSD licensed)
- portability (ISO C, POSIX, compatibility layer)
- speed and memory lighweight (compiled C binaries)
- security (replace scripts by data files, priviledge separation)
- simplicity of use (generation tools, few basic instructions)
- dependency low cost (only needs a POSIX shell and a C compiler)

The provided tools are:
- pmksetup: static data gathering tool
- pmk: dependency checker
- pmkscan: source scanner and build file generator
- pmkinstall: BSD like install program
- pmkpc: pkg-config compatible tool


1) Installation Steps
---------------------

  See the INSTALL file.


2) Testing
----------

  We provide basic tests to have some pragmatic examples of how to use
  PMK. It is also a good method to test the whole kit. You're invited to
  look in the samples directory and check the README.sample file.


3) Supported Platforms
----------------------

  Any POSIX compliant platform _should_be_ supported. We provide in the
  package a list of platform on which PMK has been compiled and tested
  (see STATUS file). You can also consult the BUGS file to see if your
  platform has specific problems that avoid PMK to run.



Changelog
27 Aug 2009 couderc
 * improved support of shared libs using versioning.

16 Jul 2009 couderc
 * completed management of dependencies between checks into generated
   pmkfile.
 * fixed missing dependency when generated check already exists.

13 Jul 2009 couderc
 * first part of modifications to handle check dependencies.

11 Mar 2009 couderc
 * unbreak library detection.

24 May 2008 coudercd
 * fixed pmkscan to check linker type values in scanfile (noticed by
   Magnus Holmgren).

11 Mar 2008 coudercd
 * fix recording of library path broken by previous commit.

10 Apr 2008 coudercd
 * use collected CFLAGS and LDFLAGS during checks (fix problems with
   hierarchical includes or libraries).

09 Apr 2008 coudercd
 * permit user data file for pmkscan.

24 Mar 2008 coudercd
 * better use of hash_get_processed() in codebuild.
 * fixed pkgconfig stuff in cases where dependencies can be missing.
 * removed use of (broken) variable substitution in pmk.conf.
 * added pkgconfig lib path detection.

18 Jul 2007 coudercd
 * moved hash_t to hash.c and done some cleaning.
 * adding dynary.c, dynary.h, dynary_tools.c and dynary_tools.h for
   further use (to replace basic dynarrays).

01 Jun 2007 coudercd
 * added support of LIBS into generated makefile templates from pmkscan.
 * added detection of sete*id() functions.

25 May 2007 coudercd
 * fixed use of LIBS.
 * pmk now check for for ~/.pmk/pmk.conf and if not exists for 
   SYSCONFDIR/pmk/pmk.conf.

23 May 2007 coudercd
 * fixed function detection in library check.
 * added math library check in pmkfile.
 * updated data/pmkscan.dat database.

12 May 2007 coudercd
 * updated pmkfile and scanfile.
 * added missing tag generation for pmk library check.

07 May 2007 coudercd
 * put linker flags like library path and/or name after objects to fix
   link resolution on platforms where order is important.
 * build library targets first.
 * update scanfile.5 manual page with DEFINE_BIN.

04 May 2007 coudercd
 * added DEFINE_BIN to allow specification of a binary name by giving
   the name of it's main source.

21 Apr 2007 coudercd
 * added hash_create_simple_autogrow() which is almost the same as
   hash_create_simple() but with autogrow enabled.

16 Apr 2007 coudercd
 * removed ending line with only one dot that makes Linux complaining,
   noticed by Magnus Holmgren.
 * added a carriage return that was missing in pmkscan log file message.
 * updated main Makefile template to fix GEN_FILES, noticed by Magnus
   Holmgren.

15 Apr 2007 coudercd
 * fixed library source labels generation.
 * fixed object dependencies when an object is used in both a binary
   and a library to avoid duplicates.
 * fixed processing of pkg-config library directory to support
   multiples path, noticed by Marco Ghirlanda.

10 Apr 2007 coudercd
 * resized ressources in pmkscan for larger projects.
 * added more messages in pmkscan.

05 Apr 2007 coudercd
 * removed obsolete option LIBOBJ.
 * added missing initialisation of lib_type in the scanzone structure.

21 Mar 2007 coudercd
 * replaced blank lines by one dot lines in manual pages, notified by
   Magnus Holmgren.
 * fixed the duplicates in macro GEN_FILES of generated makefiles, also
   notified by Magnus Holmgren.

20 Mar 2007 coudercd
 * fixed errors in manual pages, notified by Magnus Holmgren.

28 Feb 2007 coudercd
 * fixed the label of shared lib linking flags variable in the log
   message (noticed by barratis from toy'd project).

25 Feb 2007 coudercd
 * fixed a segfaulting bug in code_logger().

15 Feb 2007 coudercd
 * fixed a bug occuring in pmkpc when specifying both --cflags and
   --libs options. The result is now separated but a space.

24 Dec 2006 coudercd
 * fixed codebuild bug spotted by toyd team on linux.
 * reordered pmkcfg.sh compiler detection.

08 Dec 2006 coudercd
 * fixed install_lib target in makefile generation from pmkscan.
 * removed the compatibility option -c for install for portability.

27 Nov 2006 coudercd
 * imported new hash table implementation with double linked lists.

# PMK 0.10.0 release

18 Nov 2006 coudercd
 * added in pmkscan the code to create binary install directory in
   generated makefile template.
 * support alternative assembler in pmkcfg.sh via AS variable.
 * update Makefile.pmk template to create correctly the binary install
   directory.
 * also added in pmkscan the code for install data directory creation
   in the generated makefile template.

# snapshot 20061028

22 Oct 2006 coudercd
 * added BUILD_LIB_NAME() to replace BUILD_SHLIB_NAME() as it also build
   the static library name.
 * marked BUILD_SHLIB_NAME() as obsolete.
 * fixed manual page directories for usermode install.

22 Oct 2006 coudercd
 * added support for shared library versioning.
 * library version numbers become optional.

14 Oct 2006 coudercd
 * lib checking fix was incomplete, it should be now (noticed by Phil
   aka wintellect).
 * fixed -u option of pmksetup and added a reminder in man page to
   escape double quotes.

11 Oct 2006 coudercd
 * fixed label generation for objects.
 * reimplemented instal targets for data files and man pages.
 * updated pmkcfg.sh to sync with latest changes.
 * updated scanfile to avoid unwanted template modification.
 * updated Makefile.pmk with pmkscan generated template.

10 Oct 2006 coudercd
 * reimplemented binary targets in makefile template generated by
   pmkscan to follow the same behavior as for libraries.

09 Oct 2006 coudercd
 * recovered shared library support from my broken laptop which died 2
   seconds after the stuff was completed.
 * added missing file lang.c in the previous commit.

07 Oct 2006 coudercd
 * lib checking was broken in a previous commit, it is now fixed.

06 Oct 2006 coudercd
 * fixed coredumps in pmk_check_config() and pmk_check_pkg_config() due
   to missing initialization of the codebuild structure.

28 Sep 2006 coudercd
 * better management of the template files, now pmkscan can process
   them in only one pass.

24 Sep 2006 coudercd

func.h
/* $Id$ */

/*
 * Copyright (c) 2003-2006 Damien Couderc
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *    - Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    - Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *    - Neither the name of the copyright holder(s) nor the names of its
 *      contributors may be used to endorse or promote products derived
 *      from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

#ifndef _PMK_FUNC_H_
#define _PMK_FUNC_H_

#include "compat/pmk_stdbool.h"
#include "pmk.h"
#include "parse.h"

/*************
 * constants *
 ***********************************************************************/

/* node tokens */
#define PMK_TOK_SETNGS	1
#define PMK_TOK_DEFINE	2
#define PMK_TOK_SWITCH	3
#define PMK_TOK_IFCOND	4
#define PMK_TOK_ELCOND	5

/* special setting tokens */
#define PMK_TOK_SETVAR	10 /* set variable */
#define PMK_TOK_SETPRM	11 /* set parameter */

/* item command tokens */
/* checks */
#define PMK_TOK_CHKBIN	20
#define PMK_TOK_CHKINC	21
#define PMK_TOK_CHKLIB	22
#define PMK_TOK_CHKCFG	23
#define PMK_TOK_CHKPKG	24
#define PMK_TOK_CHKTYP	25
#define PMK_TOK_CHKVAR	26
/* lib stuff */
#define PMK_TOK_BLDLIB	30
#define PMK_TOK_BLDSLN	31


/* option keywords */
/* common */
#define KW_OPT_NAME			"NAME"
#define KW_OPT_DEFS			"DEFS"
#define KW_OPT_FUNCTION		"FUNCTION"
#define KW_OPT_HEADER		"HEADER"
#define KW_OPT_CFLAGS		"CFLAGS"
#define KW_OPT_LANG			"LANG"
#define KW_OPT_LIBS			"LIBS"
#define KW_OPT_MACRO		"MACRO"
#define KW_OPT_MAJOR		"MAJOR"
#define	KW_OPT_MEMBER		"MEMBER"
#define KW_OPT_MINOR		"MINOR"
#define KW_OPT_MODULE		"MODULE"
#define KW_OPT_SUBHDR		"SUBHDR"
#define KW_OPT_VERSION		"VERSION"
#define KW_OPT_VALUE		"VALUE"
#define KW_OPT_VARIABLE		"VARIABLE"
/* SETTINGS */
#define KW_SETNGS_GLANG		"LANG"
#define KW_SETNGS_TARGET	"TARGET"
#define KW_SETNGS_ACCOMP	"AC_COMPAT"
#define KW_SETNGS_ACCOMP	"AC_COMPAT"
#define KW_SETNGS_CCDTCT	"DETECT"
/* library name building */
#define KW_SL_VERSION		"VERSION"
#define KW_SL_STATIC		"STATIC"
#define KW_SL_SHARED		"SHARED"
/* obsolete stuff *//* XXX OBOSLETE */
#define KW_SL_VERS_NONE		"VERSION_NONE"
#define KW_SL_VERS_MAJ		"VERSION_MAJ"
#define KW_SL_VERS_FULL		"VERSION_FULL"


#define DEFINE_DEFAULT		"1"


/* error messages */
#define ERR_MSG_CC_CMD		"failed to build compiler command line."


/**********************************
 * type and structure definitions *
 ***********************************************************************/

/* structures */
typedef struct {
	char	kw[CMD_LEN];
	bool	(*fnp)(pmkcmd *, htable_t *, pmkdata *);
} cmdkw;


/**************
 * prototypes *
 ***********************************************************************/

bool	func_wrapper(prscell *, pmkdata *);
bool	process_node(prsnode *, pmkdata *);
bool	pmk_define(pmkcmd *, prsnode *, pmkdata *);
bool	pmk_target(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_ac_compat(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_settings(pmkcmd *, prsnode *, pmkdata *);
bool	pmk_ifcond(pmkcmd *, prsnode *, pmkdata *);
bool	pmk_elcond(pmkcmd *, prsnode *, pmkdata *);
bool	pmk_switches(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_binary(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_header(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_lib(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_config(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_pkg_config(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_type(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_check_variable(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_build_lib_name(pmkcmd *, htable_t *, pmkdata *);
bool	pmk_build_shlib_name(pmkcmd *, htable_t *, pmkdata *);

bool	pmk_set_parameter(pmkcmd *, prsopt *, pmkdata *);
bool	pmk_setparam_accompat(pmkcmd *, prsopt *, pmkdata *);
bool	pmk_setparam_glang(pmkcmd *, prsopt *, pmkdata *);
bool	pmk_setparam_target(pmkcmd *, prsopt *, pmkdata *);
bool	pmk_setparam_detect(pmkcmd *, prsopt *, pmkdata *);
bool	pmk_set_variable(pmkcmd *, prsopt *, pmkdata *);

#endif /* _PMK_FUNC_H_ */

/* vim: set noexpandtab tabstop=4 softtabstop=4 shiftwidth=4: */
